internal/serialization/compact_serializer.go

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+type CompactStreamSerializer struct {
+	typeToSchema         map[reflect.Type]Schema
+	typeToSerializer     map[reflect.Type]serialization.CompactSerializer
+	typeNameToSerializer map[string]serialization.CompactSerializer
+	schemaService        SchemaService
+	rabin                RabinFingerPrint
+}
+
+func NewCompactStreamSerializer(compactSerializationConfig serialization.CompactSerializationConfig) *CompactStreamSerializer {
+	typeToSchema := make(map[reflect.Type]Schema)
+	typeToSerializer := make(map[reflect.Type]serialization.CompactSerializer)
+	typeNameToSerializer := make(map[string]serialization.CompactSerializer)
+	serializers := compactSerializationConfig.Serializers()
+	for typeName, serializer := range serializers {
+		typeNameToSerializer[typeName] = serializer
+		typeToSerializer[serializer.Type()] = serializer
+	}
+	rabin := NewRabinFingerPrint()
+	rabin.Init()
+	return &CompactStreamSerializer{
+		schemaService:        *NewSchemaService(),
+		typeToSchema:         typeToSchema,
+		typeToSerializer:     typeToSerializer,
+		typeNameToSerializer: typeNameToSerializer,
+		rabin:                rabin,
+	}
+}
+
+func (CompactStreamSerializer) ID() int32 {
+	return TypeCompact
+}
+
+func (c CompactStreamSerializer) Read(input serialization.DataInput) interface{} {
+	schema := c.getOrReadSchema(input)
+	typeName := schema.TypeName()
+	serializer, ok := c.typeNameToSerializer[typeName]
+	if !ok {
+		panic("No compact serializer found for type: " + typeName)
+	}
+	reader := NewDefaultCompactReader(c, input.(*ObjectDataInput), schema)
+	return serializer.Read(reader)
+}
+
+func (c CompactStreamSerializer) Write(output serialization.DataOutput, object interface{}) {
+	t := reflect.TypeOf(object)
+	serializer := c.typeToSerializer[t]
+
+	schema, ok := c.typeToSchema[t]
+	if !ok {
+		schemaWriter := NewSchemaWriter(serializer.TypeName())
+		serializer.Write(schemaWriter, object)
+		schema = schemaWriter.Build(c.rabin)
+		c.schemaService.PutLocal(schema)
+		c.typeToSchema[t] = schema
+	}
+	output.WriteInt64(schema.ID())
+	writer := NewDefaultCompactWriter(c, output.(*PositionalObjectDataOutput), schema)
+	serializer.Write(writer, object)
+	writer.End()
+}
+
+func (c CompactStreamSerializer) IsRegisteredAsCompact(t reflect.Type) bool {
+	_, ok := c.typeToSerializer[t]
+	return ok
+}
+
+func (c CompactStreamSerializer) getOrReadSchema(input serialization.DataInput) Schema {
+	schemaId := input.ReadInt64()
+	schema, ok := c.schemaService.Get(schemaId)
+	if ok {
+		return schema
+	}
+	panic(hzerrors.NewSerializationError(fmt.Sprintf("The schema cannot be found with id: %d", schemaId),  nil))
```

Comment:
I know but in one of comments of Utku I realized that Java throws a SerializationError here. So, for consistency I did it. I think it does not matter but if there is a convention in go client code and it is to use string errors for SerializationErrors in serialization code, I can change it.

Do you think changing the message from Java is a good idea? Throwing errors with the same message can be less confusing because users can match errors of different clients. I will make the first letter lowercase for now.

Reply:
My mistake @srknzl sorry, on part-2 I understood the situation. Please feel free to dismiss my suggestion 🙏